primitives/core/src/sr25519.rs

```diff
@@ -537,32 +537,93 @@ impl CryptoType for Pair {
 pub mod vrf {
 	use super::*;
 	#[cfg(feature = "full_crypto")]
-	use crate::crypto::VrfSigner;
-	use crate::crypto::{VrfCrypto, VrfVerifier};
+	use crate::crypto::VrfSecret;
+	use crate::crypto::{VrfCrypto, VrfPublic};
 	use schnorrkel::{
 		errors::MultiSignatureStage,
 		vrf::{VRF_OUTPUT_LENGTH, VRF_PROOF_LENGTH},
 		SignatureError,
 	};
 
-	/// VRF transcript ready to be used for VRF sign/verify operations.
+	/// Transcript ready to be used for VRF related operations.
+	#[derive(Clone)]
 	pub struct VrfTranscript(pub merlin::Transcript);
 
 	impl VrfTranscript {
-		/// Build a new transcript ready to be used by a VRF signer/verifier.
+		/// Build a new transcript instance.
+		///
+		/// Each `data` element is a tuple `(domain, message)` composing the transcipt.
 		pub fn new(label: &'static [u8], data: &[(&'static [u8], &[u8])]) -> Self {
 			let mut transcript = merlin::Transcript::new(label);
 			data.iter().for_each(|(l, b)| transcript.append_message(l, b));
 			VrfTranscript(transcript)
 		}
+
+		/// Map transcript to `VrfSignData`.
+		pub fn into_sign_data(self) -> VrfSignData {
+			self.into()
+		}
+	}
+
+	/// VRF input.
+	///
+	/// Technically a transcript used by the Fiat-Shamir transform.
+	pub type VrfInput = VrfTranscript;
+
+	/// VRF input ready to be used for VRF sign and verify operations.
+	// Note: here we may potentially add a closure to check-before-sign...
```

Comment:
# Preamble

The data to be signed via the `vrf_sign` is passed within the `VrfSignData`.
This is an associated type that can store whatever the implementation (in this case sr25519) needs.

In the **current** form for sr25519 we pass:
- `transcript` : data that can be used to produce an arbitrary number of bytes via the vrf.
- `extra`: data to be signed but that does not contribute to the vrf output bytes. Used for example in the approval voting protocol [here](https://github.com/paritytech/polkadot/blob/master/node/core/approval-voting/src/criteria.rs#L330)
-  `output`: some pre-computed vrf output. In this case can save one operation during the signing operation if has been pre computed via the `make_output` method.

# Extension 

The comment about the possibility to add a closure refers to :
- how this `extra` is passed and
- if we want to perform a check before sign.

For example (in babe) the closure can call `make_bytes` to get the "score", check that the value is below the threshold and then sign if this is below (in contrast... now we sign and only after we check).

With the new keystore interface (i.e. the added `make_output`) this check before sign **now** can be done by doing something like:

```
// get vrf output bytes
let bytes = keystore.vrf_output(transcript).make_bytes();
if bytes < threshold {
    // sign if we are below the threshold
    sign = keystore.vrf_sign(transcript.into_sign_data())
}
```

With the closure I'm talking about we can perform something like

```
let sign_data = transcript.into_sign_data().with_check(|out| {
    if out.make_bytes() < threshold {
        Ok(None)
     } else {
        Err
    }
});
let sign = keystore.vrf_sign(sign_data);
```

Notice that I'm returning `Ok(None)`. Well here we can also return the optional `extra` data to be signed. So that it will also cover the polkadot approval voting use case.

What is *cool* here is that the sign data can store whatever the particular vrf requires. The keystore act just as a trampoline to the implementation. 
Is then the implementation of the particular vrf that is able to interpret and use the content of `VrfSignData`.

The advantage to use a closure to perform the *check-before-sign* is that it requires only one access to the keystore (one call) instead of 2.

We can easily introduce it in a follow-up

(sorry for the verbose explanation 😃)

Reply:
> The advantage to use a closure to perform the _check-before-sign_ is that it requires only one access to the keystore (one call) instead of 2.

Not sure that brings that much :P 

However, if you think it is some nice "sugar" that we need/want, please remove the `Note` comment and create an issue of your explanation here :D

Reply:
I've expressed my current tentative opinions on VRF interfaces in https://github.com/w3f/ring-vrf/blob/master/dleq_vrf/src/traits.rs#L101  It won't work for sr25519 but something similar & simpler works for sr25519.  In particular, the Secret and Signer are the same for sr25519, but not for ring VRFs.

These `sign_after_check` methods are useful primarily as pedagogy, meaning a dev who reads them might realize "oh I can make the input-output pair first, run business logic on them which produces this auxiliary message, and then sign the whole result".